### DISCUSSION TITLE
issue: form.reset(values) ignores undefined values

### DISCUSSION THREAD

**Discussion by andykog** [react-hook-form/react-hook-form#11752] [Q&A]
### Version Number

7.51.2

### Codesandbox/Expo snack

https://codesandbox.io/p/sandbox/react-hook-form-reset-undefined-w3p8ks

### Steps to reproduce

Initialize the form with some `values` or `defaultValues`:
```ts
const userProfileForm = useForm({
  values: { revision: 5, firstName: 'a', image: {url: 'https://image.png'} },
})
```

Assume that the data was re-fetched from server and the form needs to be reset to reflect the actual state (in this scenario the profile image was removed on another device):
```ts
userProfileForm.reset({revision: 6, firstName: 'a' }); // <- image was removed
```

**Result**: the image value wasn't reset.

**Additional info**: it used to work fine in version 7.21.0

**Workaround 1**: replace all undefined values with null when calling reset:
```ts
userProfileForm.reset(_.defaultsDeep(newUserProfile, {image: null}))
```

**Workaround 2**: call `reset` with and without arguments multiple times (you can try to press buttons in the pvovided codesandbox a few times, eventually it will reset lastname), this behaviour is inconsistent.

**[Discussion](https://github.com/orgs/react-hook-form/discussions/7589)**

### Expected behaviour

It should reset the form state completely to match the provided object without “merging” it with the previous values.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
we don't support `undefined` value, you can find similar rule for other hook such as `useController` `defaultValue` or `defaultValues`

**Comment by andykog** [comment]
@bluebill1049 Would be useful if there was a way to reset form state completely. Also I think `reset` behaviour should be consistent (see my comment about workaround 2) it shouldn't behave differently depending on how many time it was invoked.

**Reply by bluebill1049** [reply]
- `reset()` // would reset your form with defaultValues which you have supplied to useForm
- `reset({ ... })` // would reset whatever you have supplied in the reset payload

hook form supports both with or without `defaultValues`, which is convenient but you have to choose your impl here as well. I would consider to always supply reset with payload to update your form.

https://react-hook-form.com/docs/useform/reset
<img width="870" alt="Screenshot 2024-04-04 at 10 12 43 pm" src="https://github.com/react-hook-form/react-hook-form/assets/10513364/fc7ccc4b-de2d-4ed8-b301-53932e91b506">

**Reply by andykog** [reply]
@bluebill1049 neither `reset()` nor `reset({ test: 'test' })` will remove other values that already exist in the form state (e.g. `image: {}`, doesn't matter if you initialise form with or without `defaultValues`. This is shown in my codesandbox.

**Comment by AlevtinaLeytland** [comment]
Is there some decision?

**Comment by iemi111** [comment]
If you are not planning to support `undefined` then you should update your documentation to warn people from using this library with [zod](https://zod.dev/) and probably other libraries as zod uses `undefined` to check whether the data is required or not....

**Comment by GuoXiaoyang** [comment]
Can you describe why undefined values are not supported?
